lib/fs/casefs.go

```diff
@@ -0,0 +1,450 @@
+		return err
+	}
+	return f.Filesystem.Unhide(name)
+}
+
+func (f *caseFilesystem) checkCase(name string) error {
+	var err error
+	if name, err = Canonicalize(name); err != nil {
+		return err
+	}
+	// Stat is necessary for case sensitive FS, as it's then not a conflict
+	// if name is e.g. "foo" and on dir there is "Foo".
+	if _, err := f.Filesystem.Lstat(name); err != nil {
+		if IsNotExist(err) {
+			return nil
+		}
+		return err
+	}
+	return f.checkCaseExisting(name)
+}
+
+// checkCaseExisting must only be called after successfully canonicalizing and
+// stating the file.
+func (f *caseFilesystem) checkCaseExisting(name string) error {
+	realName, err := f.realCase(name)
+	if IsNotExist(err) {
+		// It did exist just before -> cache is outdated, try again
+		f.dropCache()
+		realName, err = f.realCase(name)
+	}
+	if err != nil {
+		return err
+	}
+	if realName != name {
+		return &ErrCaseConflict{name, realName}
+	}
+	return nil
+}
+
+type defaultRealCaser struct {
+	fs            Filesystem
+	caseRoot      *caseNode
+	caseCount     int
+	caseTimer     *time.Timer
+	caseTimerStop chan struct{}
+	caseMut       sync.RWMutex
+}
+
+func newDefaultRealCaser(fs Filesystem) *defaultRealCaser {
+	caser := &defaultRealCaser{
+		fs:        fs,
+		caseRoot:  &caseNode{name: "."},
+		caseTimer: time.NewTimer(0),
+	}
+	<-caser.caseTimer.C
+	return caser
+}
+
+func (r *defaultRealCaser) realCase(name string) (string, error) {
+	out := "."
+	if name == out {
+		return out, nil
+	}
+
+	r.caseMut.Lock()
+	defer func() {
+		if r.caseCount > caseMaxCachedNames {
+			select {
+			case r.caseTimerStop <- struct{}{}:
+			default:
+			}
+			r.dropCacheLocked()
+		} else if r.caseTimerStop == nil {
+			r.startCaseResetTimerLocked()
+		}
+		r.caseMut.Unlock()
+	}()
+
+	node := r.caseRoot
+	for _, comp := range strings.Split(name, string(PathSeparator)) {
+		if node.dirNames == nil {
+			// Haven't called DirNames yet
+			var err error
+			node.dirNames, err = r.fs.DirNames(out)
+			if err != nil {
+				return "", err
+			}
+			node.dirNamesLower = make([]string, len(node.dirNames))
+			for i, n := range node.dirNames {
+				node.dirNamesLower[i] = UnicodeLowercase(n)
+			}
+			node.children = make(map[string]*caseNode)
+			node.results = make(map[string]*caseNode)
+			r.caseCount += len(node.dirNames)
+		} else if child, ok := node.results[comp]; ok {
+			// Check if this exact name has been queried before to shortcut
+			node = child
+			out = filepath.Join(out, child.name)
+			continue
+		}
+		// Actually loop dirNames to search for a match
+		n, err := findCaseInsensitiveMatch(comp, node.dirNames, node.dirNamesLower)
+		if err != nil {
+			return "", err
+		}
+		child, ok := node.children[n]
+		if !ok {
+			child = &caseNode{name: n}
+		}
+		node.results[comp] = child
+		node.children[n] = child
+		node = child
+		out = filepath.Join(out, n)
+	}
+
+	return out, nil
+}
+
+func (r *defaultRealCaser) startCaseResetTimerLocked() {
```

Comment:
This thing could be slightly simplified by not using a timer and a routine for the rest, instead recording the creation time of the cache. If it's too old or too large on entry into realCase we drop the cache. The downside would be that a cache is not automatically expired when we stop calling realCase -- it would stick around until next call. A workaround for that could be a permanent ticker routine that once every few seconds checks if the cache is expired and removes it. This is similar to what happens now except there would be a permanent routine and ticker instead of the selects and stuff.

I don't think the current implementation is necessarily bad, but it is a few things to take in.

Reply:
I avoided a persistent goroutine because a filesystem never gets closed, thus we'd leak a goroutine every time a filesystem is (re-)created.

Reply:
Good point